ACM-18034: Doc `resourceRequirements` add-on configuration

dhaiducek opened the pull request:
This is an add-on configuration that goes into the add-on doc. However, for 2.13 it only applies to GRC. In 2.14, the remainder of the add-ons will implement it in [ACM-17874](https://issues.redhat.com/browse/ACM-17874).

ref: https://issues.redhat.com/browse/ACM-18034

dhaiducek committed:
Document `resourceRequirements` add-on configuration

ref: https://issues.redhat.com/browse/ACM-18034
Signed-off-by: Dale Haiducek <19750917+dhaiducek@users.noreply.github.com>

swopebe commented:
This is connected to feature work: https://issues.redhat.com/browse/ACM-17563

dhaiducek committed:
Fix the addon patch command

Signed-off-by: Dale Haiducek <19750917+dhaiducek@users.noreply.github.com>

Commit:
provided doc revisions

jc-berger commented on add-ons/configure_addons.adoc:
```diff
@@ -0,0 +1,128 @@
+[#configure-addons]
+= Configuring klusterlet add-ons
+
+In {acm-short}, you can configure the following klusterlet add-ons to improve the performance and functionality of your managed clusters:
```
Please let me know what you think about adding some context to the "benefit" or "why" users would configure kluterlet add-ons.

dhaiducek replied:
Yes, I think "performance and functionality" makes sense!

jc-berger commented on add-ons/configure_addons.adoc:
```diff
@@ -0,0 +1,128 @@
+* governance-policy-framework
+
+[#applying-specifications-addons]
+== Applying specifications to klusterlet add-ons
```
since we had two different procedures here, I re-structured the content a bit. Please let me know what you think, and if I am correct that the goal of this procedure/task is to instruct users to apply specification to their klusterlet add-ons

dhaiducek replied:
The reflow is good! But perhaps there could be three topics here to make referring directly to the `AddOnDeploymentConfig` more straightforward?
- Setting up the `AddOnDeploymentConfig` to configure klusterlet add-ons
- Configuring a klusterlet add-on for all managed clusters:
  - Configure the `ClusterManagementAddOn` with the `AddOnDeploymentConfig` (which applies to the add-on for all managed clusters attached to the hub cluster)
- Configuring a klusterlet add-on for a single managed cluster:
  - Configure the `ManagedClusterAddOn` with the `AddOnDeploymentConfig` (which applies to the add-on for the particular managed cluster attached to that namespace on the hub cluster)

jc-berger commented on add-ons/configure_addons.adoc:
```diff
@@ -0,0 +1,128 @@
+// The note here only applies to ACM 2.13 and should be removed in ACM 2.14.
+<5> *Note:* Configuring `resourceRequirements` is only available for policy add-ons. List resource requirements here to override the `resources` of the add-on workload containers. If an add-on container matches more than one of the items in the list, the last matching configuration is applied.
+<6> Replace `<workload-kind>` with the kind of workload, for example: `deployment`. Replace `<workload-name>` with the name of the workload. Replace `<container-name>` with the name of the container. 
+.. For any of these values, you can use `\*` attribute to apply the configuration to all objects managed by the add-on. For example, if you used the  `\*:*:*` attribute , it would apply the configuration  to every container of every workload kind in any add-on the configuration is attached to.
```
So, from reading various troubleshooting, I saw that \ before an asterisks in ticks, ( `*`) would prevent the rendering error. It worked the first time, but not the second time. I'll continue to look around for ways to render `*:*:*` but I had to take a break and move on, loll

dhaiducek replied:
I ran into this. I think the alternative syntax is a backtick and then a plus sign:
```
`+*:*:*+`
```
ref: https://docs.asciidoctor.org/asciidoc/latest/text/literal-monospace/

Alternatively, I think you can escape every single asterisk like:
```
`\*:\*:\*`
```
I'll note though that ASCIIDoc is not Markdown, so it won't necessarily render the same in GitHub as in staging/prod.

swopebe commented on add-ons/configure_addons.adoc:
```diff
@@ -0,0 +1,128 @@
+[#configure-addons]
```
I would like to change the filename here because "addons" means a lot right now within ACM. Can we indicate this is klusterlet addons? This will help bc in the future I will be rearranging this area.

dhaiducek replied:
So something like `configure_klusterlet_addons.adoc`, then?

Commit:
applied dev feedback

dhaiducek commented on add-ons/configure_klusterlet_addons.adoc:
```diff
@@ -0,0 +1,134 @@
+[#configure-klusterlet_addons]
```
Looks like this tag should have dashes throughout?

```suggestion
[#configure-klusterlet-addons]
```

jc-berger committed:
Update add-ons/configure_klusterlet_addons.adoc

Co-authored-by: Dale Haiducek <19750917+dhaiducek@users.noreply.github.com>

dhaiducek commented on add-ons/configure_klusterlet_addons.adoc:
```diff
@@ -0,0 +1,134 @@
+[#setting-addondeploymentconfig-klusterlet-addons]
+== Setting up the _AddOnDeploymentConfig_ to configure klusterlet add-ons
+
+When you configure the klusterlet add-ons, you can apply specifications to all the add-ons on each of your managed clusters, such as the `nodeSelector` and `tolerations`. To configure the klusterlet add-on, complete the following steps:
```
I think this should be "any of the add-ons"? (Since the add-ons are configured one at a time)
```suggestion
When you configure the klusterlet add-ons, you can apply specifications to any of the add-ons on each of your managed clusters, such as the `nodeSelector` and `tolerations`. To configure the klusterlet add-on, complete the following steps:
```

dhaiducek replied:
(I've updated this in a subsequent commit, in case it should be reverted.)

dhaiducek committed:
Dev edits

- Surround `nodeSelector` and `tolerations` in
curly braces to show they're objects
- Add line breaks to make `oc` command more
readable
- Clean up whitespace
- Minor wording changes

Signed-off-by: Dale Haiducek <19750917+dhaiducek@users.noreply.github.com>

dhaiducek commented:
FYI: I've added a dev commit adjusting some spacing, adding some YAML syntax, and minor wording updates.

I verified the `oc` commands just now, but I think it'd be best have someone from server foundation give it a quick look at this content also since it's technically their content, so getting a quick approval from them would be good to close the loop.

swopebe commented:
I will review this one and edit as needed for peer review.

swopebe commented on add-ons/configure_klusterlet_addons.adoc:
```diff
@@ -0,0 +1,148 @@
+* work-manager
+
+// This note only applies to ACM 2.13 and should be removed in ACM 2.14.
+*Note:* Configuring `resourceRequirements` is only available for the following add-ons:
```
Please change this a bit, the comment in this case is fine to keep, though we try to keep the source free of comments.

It's not a note, since it is rather important to the release. I would use _important_

Something like:

For {acm-short} version 2.13, you can only configure.... (Make it active) 

(Since I have cleaned out a lot of unnecessary versioning in the doc to simplify our lives and the doc, I will know to review this next release because it will come up in my versioning work I do each release.)

swopebe commented on add-ons/configure_klusterlet_addons.adoc:
```diff
@@ -0,0 +1,148 @@
+
+When you configure the klusterlet add-ons, you can apply specifications to any of the add-ons on each of your managed clusters, such as the `nodeSelector` and `tolerations`. To configure the klusterlet add-on, complete the following steps:
+
+.  Use the `AddonDeploymentConfig` API to create an add-on configuration in any namespace on the hub cluster.
```
extra space here

swopebe commented on add-ons/configure_klusterlet_addons.adoc:
```diff
@@ -0,0 +1,148 @@
+
+.  Use the `AddonDeploymentConfig` API to create an add-on configuration in any namespace on the hub cluster.
+
+. Create a file named `addondeploymentconfig.yaml` with the following template:
```
We usually also have a command to give them to create the file in a two-step process.

swopebe commented on add-ons/configure_klusterlet_addons.adoc:
```diff
@@ -0,0 +1,148 @@
+<2> Replace `config-namespace` with the namespace of the `AddonDeploymentConfig` that you created.
+<3> Replace `<node-selector>` with your node selector.
+<4> Replace `<tolerations>` with your tolerations.
+// The note here only applies to ACM 2.13 and should be removed in ACM 2.14.
```
Same thing here, version this and you can remove the comment. You can also remove _Note_ -- in fact, I will create an issue now for next release for us to remove this after checking in with the team.

swopebe replied:
Issue for next release is here: https://issues.redhat.com/browse/ACM-18367

swopebe commented on add-ons/configure_klusterlet_addons.adoc:
```diff
@@ -0,0 +1,148 @@
+.. For any of these values, you can use `\*` attribute to apply the configuration to all objects managed by the add-on. For example, if you used the `*:*:*` attribute , it would apply the configuration to every container of every workload kind in any add-on the configuration is attached to.
++
+A completed `AddOnDeployment` resembles the following example: 
++
```
Needs extra line before (+)

swopebe commented on add-ons/configure_klusterlet_addons.adoc:
```diff
@@ -0,0 +1,148 @@
+[#configuring-klusterlet-addons-all-managedclusters]
+== Configuring a klusterlet add-on for all managed clusters
+
+After you set up the `AddOnDeploymentConfig`, you can configure it with the `ClusterManagementAddOn` which then applies this add-on configuration to all your managed clusters attached to the hub cluster. To configure a klusterlet add-on for all managed clusters, complete the following steps:
```
"managed clusters attached"

Anytime we see this, we need to add "that are" 

managed clusters **that are** attached...

See the [standard.](https://www.ibm.com/docs/en/ibm-style?topic=grammar-pronouns#relative-pronouns)

swopebe commented on add-ons/configure_klusterlet_addons.adoc:
```diff
@@ -0,0 +1,148 @@
+
+After you set up the `AddOnDeploymentConfig`, you can configure it with the `ClusterManagementAddOn` which then applies this add-on configuration to all your managed clusters attached to the hub cluster. To configure a klusterlet add-on for all managed clusters, complete the following steps:
+
+. Apply the `AddOnDeploymentConfig` file to your klusterlet add-on by running the following command:
```
Need the extra line here.

swopebe commented on add-ons/configure_klusterlet_addons.adoc:
```diff
@@ -0,0 +1,148 @@
+----
+
+. Use the configuration that you created as the global default configuration for your add-on by running the following command:
++
```
same

swopebe replied:
We need to tell the user that they are patching something here @dhaiducek -- some more detail; from the line before, we just tell them to use the configuration and run a command.

swopebe commented on add-ons/configure_klusterlet_addons.adoc:
```diff
@@ -0,0 +1,148 @@
+. Use the `AddonDeploymentConfig` API to create another configuration to specify the `nodeSelector` and `tolerations` on the hub cluster.
+
+. Connect the new configuration that you created to your `ManagedClusterAddon` add-on on the managed cluster by running the following command:
++
```
Need extra space and a little more detail about the "patch"?

swopebe reviewed (changes requested):
this looks really good, but there are a few comments, changes required, and suggestions. I will create the issue so that we can remove the comments.

Commit:
doc feedback

dhaiducek committed:
Add context around the `oc patch` cmd

Signed-off-by: Dale Haiducek <19750917+dhaiducek@users.noreply.github.com>

jc-berger reviewed (approved):
/lgtm
